hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/AuditTool.java

```diff
@@ -0,0 +1,308 @@
+   * @return SUCCESS i.e, '0', which is an exit code.
+   * @throws Exception on any failure.
+   */
+  @Override
+  public int run(String[] args) throws Exception {
+    List<String> argv = new ArrayList<>(Arrays.asList(args));
+    if (argv.isEmpty()) {
+      errorln(getUsage());
+      throw invalidArgs("No bucket specified");
+    }
+    //Path of audit log files in s3 bucket
+    Path s3LogsPath = new Path(argv.get(0));
+
+    //Setting the file system
+    URI fsURI = toUri(String.valueOf(s3LogsPath));
+    S3AFileSystem s3AFileSystem =
+        bindFilesystem(FileSystem.newInstance(fsURI, getConf()));
+    RemoteIterator<LocatedFileStatus> listOfS3LogFiles =
+        s3AFileSystem.listFiles(s3LogsPath, true);
+
+    //Merging local audit files into a single file
+    File s3aLogsDirectory = new File(s3LogsPath.getName());
+    boolean s3aLogsDirectoryCreation = false;
+    if (!s3aLogsDirectory.exists()) {
+      s3aLogsDirectoryCreation = s3aLogsDirectory.mkdir();
+    }
+    if(s3aLogsDirectoryCreation) {
+      while (listOfS3LogFiles.hasNext()) {
+        Path s3LogFilePath = listOfS3LogFiles.next().getPath();
+        File s3LogLocalFilePath =
+            new File(s3aLogsDirectory, s3LogFilePath.getName());
+        boolean localFileCreation = s3LogLocalFilePath.createNewFile();
+        if (localFileCreation) {
+          FileStatus fileStatus = s3AFileSystem.getFileStatus(s3LogFilePath);
+          long s3LogFileLength = fileStatus.getLen();
+          //Reads s3 file data into byte buffer
+          byte[] s3LogDataBuffer =
+              readDataset(s3AFileSystem, s3LogFilePath, (int) s3LogFileLength);
+          //Writes byte array into local file
+          FileUtils.writeByteArrayToFile(s3LogLocalFilePath, s3LogDataBuffer);
+        }
+      }
+    }
+
+    //Calls S3AAuditLogMerger for implementing merging code
+    //by passing local audit log files directory which are copied from s3 bucket
+    S3AAuditLogMerger s3AAuditLogMerger = new S3AAuditLogMerger();
+    s3AAuditLogMerger.mergeFiles(s3aLogsDirectory.getPath());
+
+    //Deleting the local log files
+    if (s3aLogsDirectory.exists()) {
+      FileUtils.forceDeleteOnExit(s3aLogsDirectory);
+    }
+    return SUCCESS;
+  }
+
+  /**
+   * Read the file and convert to a byte dataset.
+   * This implements readfully internally, so that it will read.
+   * in the file without ever having to seek().
+   *
+   * @param s3AFileSystem   filesystem.
+   * @param s3LogFilePath   path to read from.
+   * @param s3LogFileLength length of data to read.
+   * @return the bytes.
+   * @throws IOException IO problems.
+   */
+  private byte[] readDataset(FileSystem s3AFileSystem, Path s3LogFilePath,
+      int s3LogFileLength) throws IOException {
+    byte[] s3LogDataBuffer = new byte[s3LogFileLength];
+    int offset = 0;
+    int nread = 0;
+    try (FSDataInputStream fsDataInputStream = s3AFileSystem.open(
+        s3LogFilePath)) {
+      while (nread < s3LogFileLength) {
```

Comment:
* best to read in say 2MB byte array and save to output stream passed in
* you can use readFully, but this doesn't actually handle that final block of data.